### ISSUE TITLE
Investigate providing `&mut world` in tweenable callback vs. using events

### ISSUE THREAD

**Issue by djeedai** [djeedai/bevy_tweening#3]
Currently a tweenable callback has little information. It contains a reference to the `Tween<T>`, and (change not merged yet) the `Entity` of the `Animator<T>` or `AssetAnimator<T>`.

Kero on Discord suggested providing a `&mut world` to allow the callback to do something useful, similar to [what `bevy-tick-timers` does](https://github.com/maplant/bevy-tick-timers/blob/master/example/src/main.rs). But this forces the system ticking the animators to take a global lock and become exclusive, serializing execution and being a potential choke point for other Bevy systems, which can cause performance issues. This is particularly bad for `bevy_tweening` which has a series of such systems, once per component type:

https://github.com/djeedai/bevy_tweening/blob/38e99394749d834f5afd7e5bf76b4692b7d44ce1/src/plugin.rs#L48-L59

A possible alternative which sounds more Bevy-esque would be to replace the callback with a Bevy event. This way the user app can design any system they want to read that event, with any resource / query they need. This however implies being able to write an event without knowing its type at build time, to avoid depending on the event type as an extra generic parameter for the system:

```rust
let mut tween = Tween::new(...);
tween.set_completed::<MyEvent>();
```

then

```rust
pub fn component_animator_system<T: Component>( 
    time: Res<Time>, 
    mut query: Query<(&mut T, &mut Animator<T>)>,
    mut writer: EventWriter<????>, //< Here we don't know the type
) { 
    for (ref mut target, ref mut animator) in query.iter_mut() { 
        if animator.state != AnimatorState::Paused { 
            if let Some(tweenable) = animator.tweenable_mut() { 
                tweenable.tick(time.delta(), target, writer); 
            } 
        } 
    } 
} 
```

**Comment by alice-i-cecile** [comment]
As a simpler alternative, create `Events<TweenCompleted>` (and so on) event streams, which stores the `Entity` that the tween is associated with along with other alternative information.

Then, users can follow that event stream, and respond to the animation events in their own systems.

**Comment by Kerollmops** [comment]
Just found out that it would maybe be better to insert a component to an entity and remove it when the animation is finished. [It is the way benimator does it](https://docs.rs/benimator/latest/benimator/#run-the-animation-only-once), it inserts the Play component to the animated entity (Sprite) and removes it when the animation was set to `Once` and is finished.

One can detect the removal of a component pretty easily by using the `RemovedComponent` special system parameter, the only "tricky" part would be on this crate side, [you can read more on the Bevy cheat sheet](https://bevy-cheatbook.github.io/programming/removal-detection.html?highlight=removed#removal-detection).

**Comment by djeedai** [comment]
Support has been merged via e31e76dcfa5d2d7973431ad5d110575d2c63eee1 and 8dcd3d27e9e516efc4672dae912741953d08d192, so closing this. The `sequence` example provides a use case example, which given how easy it is to use feels like the approach is superior to add/remove components. We can revisit the question of `&mut world` and the dynamic event type if usage proves them useful.
